src/ripple/app/tx/impl/CheckAndConsumeTicket.cpp

```diff
@@ -0,0 +1,197 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/tx/impl/CheckAndConsumeTicket.h>
+#include <ripple/app/tx/TransactionEngine.h>              // TransactionEngine
+#include <ripple/app/ledger/MetaView.h>                   // MetaView
+#include <ripple/protocol/Indexes.h>                      // getOwnerDirIndex
+#include <ripple/protocol/STTx.h>                         // STTx
+
+
+namespace ripple {
+
+namespace
+{
+
+struct TicketContents
+{
+    AccountID owner;
+    std::uint32_t seq;
+
+    TicketContents () = delete;
+    TicketContents (AccountID owner_, std::uint32_t seq_)
+    : owner (owner_)
+    , seq (seq_)
+    { }
+};
+
+TicketContents getTicketContents (STTx const& txn)
+{
+    auto const& ticketID = txn.getFieldObject (sfTicketID);
+
+    return {ticketID.getAccountID (sfAccount),
+        ticketID.getFieldU32 (sfSequence)};
+}
+
+// Return the appropriate error code for this missing Ticket.
+TER missingTicketTER (TicketContents const& ticket, SLE::ref sleOwner)
+{
+    // If the account owner can't be found give up.
+    if (! sleOwner)
+        return tefNO_ENTRY;
+
+    // 1. If the Ticket sequence exceeds the ticket owner's seq, then the
+    //    Ticket may not have been created in this ledger yet.  Retry.
+    //
+    // 2. If the Ticket sequence is the same or less than the ticket owner's
+    //    seq, then the Ticket was either never created or was consumed.  Fail.
+    auto const ownerSeq = sleOwner->getFieldU32 (sfSequence);
+    return ticket.seq >= ownerSeq ? terPRE_TICKET : tefNO_ENTRY;
+}
+
+bool authorizedToUseTicket (SLE::ref sleTicket,
+    AccountID const& txnAccountID, AccountID const& ticketOwner)
+{
+    if (txnAccountID == ticketOwner)
+        return true;
+
+    // The target can also always consume a ticket.
+    if (sleTicket->isFieldPresent (sfTarget))
+        return (txnAccountID == sleTicket->getAccountID (sfTarget));
+
+    return false;
+}
+
+bool expiredTicket (SLE::ref sleTicket, TransactionEngine* const txnEngine)
+{
+    if (sleTicket->isFieldPresent (sfExpiration))
+    {
+        std::uint32_t const expiration = sleTicket->getFieldU32 (sfExpiration);
+
+        if (txnEngine->getLedger()->getParentCloseTimeNC() >= expiration)
+            return true;
+    }
```

Comment:
I can't make that change yet, since `GetParentCloseTimeNC()` is not yet available on `BasicView`.  Perhaps with the next rebase :-).